library/qtest/src/Functions.qs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/// # Summary
+/// Runs a series of named test functions with expected results.
+///
+/// # Description
+/// Given an array of test names, test functions, and expected results (of the same type), returns `true`
+/// if all tests passed, and `false` if any one test failed.
+/// Prints the expected and received result for any failed tests.
+///
+/// # Input
+/// ## test_cases
+/// An array of three-arity tuples of the form `(test_name, callable_to_test, expected_result)`.
+/// `callable_to_test` will be called and its result will be compared to `expected_result`.
+///
+/// # Example
+/// ```qsharp
+/// function Main() : Unit {
+///     TestCases([
+///         ("Should return 42", TestCaseOne, 42),
+///         ("Should add one", () -> AddOne(5), 6)
+///     ]);
+/// }
+///
+/// function TestCaseOne() : Int {
+///     42
+/// }
+///
+/// function AddOne(x: Int) : Int {
+///     x + 1
+/// }
+/// ```
+function TestCases<'Result : Eq + Show > (test_cases : (String, () -> 'Result, 'Result)[]) : Bool {
+    let failed_test_buf = TestCasesSilent(test_cases);
+
+    if Length(failed_test_buf) == 0 {
+        Message($"{Length(test_cases)} test(s) passed.");
+        true
+    } else {
+        Message($"{Length(failed_test_buf)} tests failed.");
+        for item in failed_test_buf {
+            Message($"{item}")
+        }
+        false
+    }
+}
+
+/// # Summary
+/// Similar to `Qtest.Functions.TestCases`, but returns test failure info in an array of strings instead of directly messaging to
+/// output. Useful if you want to handle your own messaging when writing tests, for CI or similar.
+
+/// See `Qtest.Functions.TestCases` for more details.
+///
+/// # Description
+/// Given an array of test names, test functions, and expected results (of the same type), returns an array
+/// of strings representing all failed test cases (if any). Strings are of the form "test_name: expected {}, got {}"
+///
+/// # Input
+/// ## test_cases
+/// An array of three-arity tuples of the form `(test_name, callable_to_test, expected_result)`.
+/// `callable_to_test` will be called and its result will be compared to `expected_result`.
+///
+/// # Example
+/// ```qsharp
+/// function Main() : Unit {
+///     let failure_messages = TestCasesSilent([
+///         ("Should return 42", TestCaseOne, 42),
+///         ("Should add one", () -> AddOne(5), 6)
```

Comment:
Question: If someone wants to test AddOne function for multiple arguments, will they have multiple entries in the TestCases array? I was wondering if it could be simplified by having input array and expected output array for a function? This is motivated by pytest.paramererize that allows listiing inputs and expected outputs

Reply:
I think that could be another really good bit of functionality, perhaps as a follow-on addition to this library? I like how that looks, but I don't think it blocks this PR -- what do you think? I'd be happy to take on the task or review your code if you wanted to take a swing at it.

Reply:
#2007 is in now, so on the latest build these class constraints work. You'll have to build from main until we do a release, though (which should be soon).

Reply:
Sure sounds good. Created#2026 for the same.